### DISCUSSION TITLE
Updated developer documentation

### DISCUSSION THREAD

**Discussion by justinclift** [getredash/redash#6129] [Announcements]
**2024.09.10 update** - There's a new set of wiki pages showing the database structure and relationship between the tables:  https://github.com/getredash/redash/wiki#redash-database-structure

Also, I think we'll need to figure out a better way to notify people of new developer oriented content too. :wink:

**2024.09.05 update** - And there's now a [local development environment setup guide for macOS](https://github.com/getredash/redash/wiki/Local-development-setup-%E2%80%90-macOS-Sonoma-(14.6)) too.

**2024.08.01 update** - There's now a more streamlined version of the local developer setup guide [here](https://github.com/getredash/redash/wiki/Local-development-setup-%E2%80%90-Debian-12), this time based upon Debian 12 and using [asdf](https://asdf-vm.com) for installing Python + NodeJS.

---

Finally got around to writing up a new "How to set up your development environment" page today on the wiki, that works for the current Redash git repo:

https://github.com/getredash/redash/wiki/Local-development-setup

It's been written while using an Ubuntu 20.04 desktop VM, and I've rolled back the VM and gone through these steps several times now (refining things as I go).

So it's in a "known good" state for the current state of the Redash git repo.

**Comment by gaecoli** [comment]
> Finally got around to writing up a new "How to set up your development environment" page today on the wiki, that works for the current Redash git repo:
> 
> https://github.com/getredash/redash/wiki/Local-development-setup
> 
> It's been written while using an Ubuntu 20.04 desktop VM, and I've rolled back the VM and gone through these steps several times now (refining things as I go).
> 
> So it's in a "known good" state for the current state of the Redash git repo.
> 
> That being said, I'm still adding more stuff to the end of that document for reference purposes. So don't be overly worried about a bunch of TODO items at the end of it. 😄

I didn't find the installation method under mac, do we need support?

**Reply by justinclift** [reply]
Oh interesting idea.  Yeah, if you want to write the same kind of thing (but for macOS), that would be really welcome. :smile:

**Reply by justinclift** [reply]
The section of the page that I've been adding to the end, where it has info about all of the "make" targets... that info is independent of OS.  So I've moved that section to [a separate page](https://github.com/getredash/redash/wiki/The-list-of-Redash-make-commands) so it's not duplicated. :smile:

**Reply by justinclift** [reply]
As a data point, the `preview` image on Docker Hub now support ARM64 (again?).

So getting things happening on macOS again might now work. :smile:

**Comment by stephendwolff** [comment]
I have been trying the setup today, but not having much luck. I opened an issue: #6973

You said above:

> So it's in a "known good" state for the current state of the Redash git repo.

Could you provide the commit where it was known to work (on Ubuntu 20.04LTS)?

**Reply by justinclift** [reply]
It was working when I personally last checked about... um... 2 weeks ago I think.

From looking at the issue you filed, it seems like something is going wrong with the automatic database initialisation bit now.  I'll have to take a look into that later on today to find out why it's breaking, and fix it.

Thanks for opening that issue and letting us know there's a problem. :smile:

**Reply by justinclift** [reply]
In the meantime, @eradman has updated the docs to fix things and get them back to a working state. :smile:

**Reply by stephendwolff** [reply]
@justinclift ok thanks - i was going to submit a documentation PR at some point when i'm back on my `redash` work (1st day was yesterday). I did wonder whether the auto database init didn't happen as i plowed ahead after `yarn` didn't work (as i thought perhaps i could just get the backend running before solving that). But it might be unrelated!

**Comment by jpspringall** [comment]
Hi @justinclift

I've been running through the documentation to get setup and running and all good on WSL2

Now I've moved on attempting to attempting to debug ReDash

I've found this: https://redash.io/help/open-source/dev-guide/debugging/ which works well with regards to debugging the server 

However as most of the heavy lifting goes on in the worker, I'm just wondering if there is any guidance on how best to do this?

Cheers

James

**Reply by justinclift** [reply]
Really good question.  I personally use a different IDE (JetBrains IntelliJ) for my development activities, though I've recently started trying out VS Code for some stuff too.

Haven't tried it with Redash though.  Looking at those instructions, my best guess (without trying the above instructions yet) is that you'll need to start the `worker` container with the debug option and service ports, *after* starting the `server` container normally.

So, something like this to initialise the internal PostgreSQL database and start Redis:

```
$ make create_database
```

Then start the `server` container as well:

```
$ docker compose up -d server
[+] Running 3/3
 ✔ Container redash-postgres-1  Running                                                                                                                                                       0.0s 
 ✔ Container redash-redis-1     Running                                                                                                                                                       0.0s 
 ✔ Container redash-server-1    Started
```

Verify those three containers are running:

```
$  docker compose ps
NAME                IMAGE                                COMMAND                  SERVICE    CREATED          STATUS          PORTS
redash-postgres-1   pgautoupgrade/pgautoupgrade:latest   "/usr/local/bin/dock…"   postgres   10 seconds ago   Up 10 seconds   0.0.0.0:15432->5432/tcp, :::15432->5432/tcp
redash-redis-1      redis:7-alpine                       "docker-entrypoint.s…"   redis      10 seconds ago   Up 10 seconds   6379/tcp
redash-server-1     redash-server                        "/app/bin/docker-ent…"   server     10 seconds ago   Up 9 seconds    0.0.0.0:5678->5678/tcp, :::5678->5678/tcp, 0.0.0.0:5001->5000/tcp, :::5001->5000/tcp
```

Then start the `worker` container ready for debugging:

```
$ docker compose run --rm --service-ports worker debug
```

Actually, @eradman, @wlach, or @gaecoli might use VS Code and know exactly how to get it running with that? :smile:

**Reply by lucydodo** [reply]
You can also open the built-in terminal in VSCode and run those commands. That's how I'm working with it.

**Reply by jpspringall** [reply]
Hi @justinclift,

I'm really not sentimental about IDEs so I'm happy to use IntelliJ if it's easier, though from memory there were debugging limitations with the community edition?

Looking at the docker-entrypoint, it looks like the debug flag only applies to the server and not the worker: https://github.com/getredash/redash/blob/master/bin/docker-entrypoint#L120-L123

and it's  export REMOTE_DEBUG=1 that enables the debugging in https://github.com/getredash/redash/blob/master/redash/__init__.py#L20-L23

So I had tried running `docker compose run --rm --service-ports worker debug` but obviously the debugger isn't attaching

Apologies as it feels like I'm missing something really obvious!

However, @lucydodo you may be able to save the day....

Would it be possible for you to share your launch.json? As that may explain why debugger isn't attaching to the worker?

Cheers

James

**Reply by justinclift** [reply]
@jpspringall Might be useful to keep an eye on this, as I'm now going through the process to get the debugging of these bits all worked out (and will update/write docs as needed): https://github.com/getredash/redash/pull/7110

**Reply by jpspringall** [reply]
@justinclift Awesome, if you want any feedback, or want someone test the instructions, then feel free to give me a shout

**Reply by justinclift** [reply]
Definitely will do. :smile:

**Comment by eradman** [comment]
Next step is to support Ubuntu 24 dev environment. Closing this discussion
